### ISSUE TITLE
How to find the optimal k?

### ISSUE THREAD

**Issue by yang-william** [scverse/scanpy#25]
is there a way to systematically find the ideal k? There is a find_sigmas function in the R destiny package. Is the parameter k related to the parameter sigma? Thanks.

**Comment by falexwolf** [comment]
No, there is no way within Scanpy. I'll talk to Philipp about the find_sigmas function... and get back to you.

My personal opinion is that in a wide range of values, the qualitative (significant) results should be independent of the value of k.

The default value of k=30, meaning that we construct a k-nearest neighbor graph in which each cell is connected with 30 neighbors, yields good results on all data sets (>10) that I worked with so far. If you have very little noise, for example, by selecting only very few highly variable genes in the preprocessing, you might obtain a more "pronounced structure" by reducing k (I'd recommend at least 3, though).

Also with very noisy data, k=30 should be high enough to average out noise effects.

To summarize, k=30 is a conservative choice that in my experience does the job for everything. In some cases, it pays off to reduce the value.

**Comment by yang-william** [comment]
Cool thanks! Also, are there any ways to combine multiple adatas or read multiple csv files?

**Comment by falexwolf** [comment]
No, not right now. You have to do this with numpy or pandas. adata is just a collection of numpy arrays and a dict (`print(adata)`) So you can simply do this manually.

If you explain in a more detailed way what you want, I can probably also quickly implement it.

**Comment by yang-william** [comment]
Basically, I have multiple CSV files, and I would like to run dpt on all of the csv file combined. At the moment, I'm manually copying them and feeding it into a new CSV. However, that is very time consuming and my data is quite large. So, I'm wondering if there is a way to systematically handle multiple CSV files and run a single dpt on all of them. Thanks.

**Comment by yang-william** [comment]
Also, there the parameter n_dcs is not documented in the dpt script. Will changing this parameter yield better results? Or should I just stay with the default 10?

**Comment by falexwolf** [comment]
Regarding `n_dcs`. It's not a "user parameter", just stick with the default value. Changing it reasonably won't have any effect, is my current take on it. I've never done production calculation with another value.

Regarding multiple csv files: each of these files is a n_sample x n_genes table? Or transposed? I could easily write something that concatenates these tables, sure! I hope I find time for it within the next 7 days or so.

**Comment by falexwolf** [comment]
Just in general: the main route to "improving the results" is to make sure that your preprocessing yields a meaningful tSNE for the standard parameters. Then also DPT will perfectly do its job.

**Comment by yang-william** [comment]
Okay, so run tSNE first. If it yields meaningful results, then DPT will do fine? So, we don't need to play around with the DPT parameters at all? By preprocessing, you mean like gene filtering, normalization, etc?

Also, the CSV files are in n_sample x n_genes, but ideally, it would be great if we could combine multiple adata together, so we can transpose it ourselves (and other preprocessing), if we need to. Thanks!

P.S
I'm currently running DPT with n_pcs=0, so no dimension reduction with PCA. Not sure if this is good or not.

**Comment by falexwolf** [comment]
You don't need to play around with the DPT parameters at all. You should definitely run it with `n_pcs = 50` (the default value). In some cases, it pays off reducing this. But in almost no cases (except if you have very low dimensional data, for example, from qPCR) it's meaningful to set `n_pcs = 0`.

By preprocessing I mean gene filtering, batch correction and normalization (e.g. `sc.pp.recipe_zheng17(adata)` does everything, or see the Seurat example).

**Comment by yang-william** [comment]
Thank you very much for the help!

Sorry to bother you further, but if I'm understanding filter_gene_dispersion, it is filtering out the highly varying genes. Why would you want to do that? Don't you want to remove genes that doesn't vary and keep the ones that vary a lot because you would think they play a role in cell differentiation? 

Also, I don't have much experience with t-SNE. How can we tell if our t-SNE is meaningful? 

For example, this data set has an excellent DPT result
t-sne
![tsne2](https://user-images.githubusercontent.com/23245419/27449211-8e39fa20-5755-11e7-9ef2-175626549800.png)
dpt
![dpt2](https://user-images.githubusercontent.com/23245419/27449233-9cfbc3fe-5755-11e7-84f8-9e98d88308e7.png)

meanwhile this data set has no branches at all

tsne
![tsne](https://user-images.githubusercontent.com/23245419/27449514-abdd1192-5756-11e7-956e-0a702bb51fa4.png)
dpt
![dpt](https://user-images.githubusercontent.com/23245419/27449525-b35f817a-5756-11e7-9499-4b3a79d61dfb.png)


Is there anything in particular I should look for? Thanks!

P.S
my dpt runs faster than tSNE

**Comment by falexwolf** [comment]
The tSNE looks "good" if parts of the data that you expect to be connected actually look connected in the tSNE and do not cluster apart. In your case, it looks a bit too clustered, but seeing the DiffMaps, everything turns out to be fine. The second example simply doesn't seem to have a branching.
PS: There will soon be a new default tSNE that will ensure that the correspondence between DiffMap and tSNE is better. Still, of course, both visualization methods focus on different aspects of the data.

**Comment by yang-william** [comment]
Is there any particular reason why DPT fails when you don't have enough genes? I get this error when I select 23 genes for use

```
scipy.sparse.linalg.eigen.arpack.arpack.ArpackError: ARPACK error 3:
No shifts could be applied during a cycle of the Implicitly restarted Arnoldi iteration.
One possibility is to increase the size of NCV relative to NEV.
```

with a lot of warnings.

Also, is there any particular reason why on some datasets, you can't find significant genes via filter_gene_dispersion? Does that mean that dataset is bad?

**Comment by falexwolf** [comment]
No, there should not be any reason that is associated with a small number of genes per se. In the moignard15 example, everything works for 40 genes; in the toggleswitch, everything works for 2 genes.

Does your PCA look meaningful? Try supplying a very small number of PCs to DPT (`n_pcs=3` or so).

If you do not find significant genes with `filter_genes_dispersion`, you have to adapt the parameters [e.g. set `min_disp` to a lower value](https://github.com/theislab/scanpy/blob/2cea8341e28eb8d0658f62d010631f77465e16d7/scanpy/preprocessing/simple.py#L132-L177). See the example [here](https://github.com/theislab/scanpy_usage/blob/master/170505_seurat/seurat.ipynb). Alternatively, you can simply select the `n_top_genes` highest variabale genes by setting `flavor` to `'cell_ranger'`.
